File: Mail.jl
module Mail

export sendmail

using SMTPClient
using Dates
using Base64

const bset = [collect('A':'Z'); collect('a':'z'); collect('0':'9')]

function sendmail(;from=(name=nothing, mail=nothing),
                   tolist=[(name=nothing, mail=nothing)],
                   subject="", message="",
                   attachfiles=nothing, url="smtps://smtp.gmail.com:465")

  ret = 0

  secret = Base.getpass("Password")
  passwd = string(Char.(secret.data)...)
  Base.shred!(secret)

  for to in tolist
    date = "Date: "*Dates.format(now(),"e, d u Y HH:MM:SS")*" +0530 (IST)\r\n"
    opt = SendOptions(isSSL = true, username = from.mail, passwd = passwd)
    boundary = string(rand(bset,20)...)
    sub = replace(subject, r"(\$receiver)"=>to.name)
    msg = replace(message, r"(\$receiver)"=>to.name)
    msg = replace(msg, r"(\$subject)"=>sub)

    attach = ""
    if attachfiles !== nothing
      for afile in attachfiles
        attachment = base64encode(read(afile))
        attach *= "--$boundary\r\n" *
                  "Content-Type: application/octet-stream\r\n" *
                  "Content-Transfer-Encoding: base64\r\n" *
                  "Content-Disposition: attachment; filename=$afile\r\n" *
                  "\r\n" *
                  "$attachment" * "\r\n"
      end
    end

    body = IOBuffer(date * "From: $(from.name) <$(from.mail)>\r\n" *
                           "To: $(to.name) <$(to.mail)>\r\n" *
                           "Subject: $sub\r\n" *
                           "Content-Type: multipart/mixed; boundary=$boundary\r\n" *
                           "MIME-Version: 1.0\r\n" *
                           "--$boundary\r\n" *
                           "Content-Type: text/html; charset=\"us-ascii\"" *
                           "\r\n" *
                           msg * "\r\n" *
                           attach *
                           "--$boundary--" * "\r\n"
                   )
    
    try
      resp = send(url, [to.mail], from.mail, body, opt)
      println("Sending $subject to $(to.mail)")
      ret = 1
    catch
      ret = 0
    end
  end

  if ret == 0
    println("Incorrect password. Try again.")
  end
  return ret

end

end

File: mail.py
import email, smtplib, ssl, getpass
import smtplib
from email.mime.multipart import MIMEMultipart
from email.mime.text import MIMEText
from email.mime.base import MIMEBase
from email import encoders

def get_message(sender_email, sender_name, emails, receiver_names, fname, subject):
    # Create a multipart message and set headers
    body = "Dear " + receiver_names + ",<br><br>" + f"""\
            Please find attached the {subject} for encryption
            """
    message = MIMEMultipart()
    message["From"] =  sender_name+"<"+sender_email+">"
    message["To"] = receiver_names + "<" + emails + ">"
    message["Subject"] = subject
    # message["Bcc"] = "krishvi-dhavala-committee@googlegroups.com"

    # Add body to email
    message.attach(MIMEText(body, "html"))

    # In same directory as script
    files = [fname]

    for f in files:
        # Add file as application/octet-stream
        # Email client can usually download this automatically as attachment
        part = MIMEBase('application', "octet-stream")
        # Open PDF file in binary mode
        part.set_payload(open(f, "rb").read())
        # Encode file in ASCII characters to send by email
        encoders.encode_base64(part)
        part.add_header("Content-Disposition", f"attachment; filename= {f}",)
        # Add attachment to message
        message.attach(part)

    # Convert message to string
    return message.as_string()

def mailto(sender_email, sender_name, receiver_emails, receiver_names, fname, subject):
    #sender_email = "achyub@gmail.com"
    #receiver_emails = "achyut.22068@gear.ac.in"
    password = getpass.getpass("Type your password and press enter:")
    # Log in to server using secure context and send email
    context = ssl.create_default_context()
    with smtplib.SMTP_SSL("smtp.gmail.com", 465, context=context) as server:
        server.login(sender_email, password)
        #now = datetime.now().strftime("%m-%d-%Y-%H:%M:%S")
        emails = receiver_emails.split(',')
        print(f"Sending email with email ids: {receiver_emails}")
        for i in range(0,len(emails)):
          text = get_message(sender_email, sender_name, emails[i], receiver_names[i], fname, subject)
          server.sendmail(sender_email, emails[i], text)
